crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -32,13 +39,20 @@ pub struct SemanticSyntaxChecker {
     /// Python considers it a syntax error to import from `__future__` after any other
     /// non-`__future__`-importing statements.
     seen_futures_boundary: bool,
+
+    /// The checker is currently in an `async` context: either the body of an `async` function or an
+    /// `async` comprehension.
+    ///
+    /// Note that this should be updated *after* checking the current statement or expression
+    /// because the parent context is what matters.
+    in_async_context: bool,
+
+    checkpoint: Vec<Checkpoint>,
```

Comment:
I wonder if we could avoid having to store the checkpoints inside and instead move it to the stack by changing `SemanticSyntaxChecker::enter_stmt` to return a state which the caller needs to pass to `exit_stmt` (consumes the state, the state can't be created externally).

The checker would always store the current state and `enter_stmt` would return the *previous* state from `enter_stmt` (but update its internal state). The `exit_stmt` call (or expression) then restores the internal state.

The downside of this is that it requires some more bookkeeping at the call site, but it seems fairly reasonable.

This approach would mirror the Checker's approach with updating `SemanticModelFlags`, where it updates the flags before walking the children and then restores the flags once it's done.

Reply:
If you decide to keep the stack solution, we should then change the implementation to only snapshot and restore when necessary. For example, it isn't necessary to store the snapshot for most expressions because the state won't be changed (and, therefore, there isn't anything to restore either).